README.md
Marxco Iconfont CS
---

[![GitHub package.json version (branch)](https://img.shields.io/github/package-json/v/m-o-z-g/marxco-icf/master?label=version&logo=github)](https://github.com/M-O-Z-G/marxco-icf) [![GitHub Release Date](https://img.shields.io/github/release-date-pre/m-o-z-g/marxco-icf?logo=github)](https://github.com/M-O-Z-G/marxco-icf) [![Icons Count](https://frama.link/xHxyshm8)](https://m-o-z-g.gitlab.io/marxco-icf/) [![pipeline status](https://img.shields.io/gitlab/pipeline/m-o-z-g/marxco-icf?logo=gitlab&label=pipeline)](https://gitlab.com/M-O-Z-G/marxco-icf/pipelines) [![coverage report](https://gitlab.com/M-O-Z-G/marxco-icf/badges/master/coverage.svg)](https://m-o-z-g.gitlab.io/marxco-icf/coverage) [![AppVeyor](https://img.shields.io/appveyor/build/m-o-z-g/marxco-icf?logo=appveyor&logoColor=white)](https://ci.appveyor.com/project/M-O-Z-G/marxco-icf) [![devDependency Status](https://david-dm.org/m-o-z-g/marxco-icf/dev-status.png)](https://david-dm.org/m-o-z-g/marxco-icf?type=dev) [![GitHub package.json dependency Gulp version (dev dep on branch)](https://img.shields.io/github/package-json/dependency-version/m-o-z-g/marxco-icf/dev/gulp?logo=gulp&logoColor=white&label=Gulp)](https://github.com/gulpjs/gulp) [![node](https://img.shields.io/static/v1?label=Node.js&message=%3E=8.17.0&logo=node.js&logoColor=white&color=informational)](https://github.com/nodejs/node) [![Activity](https://img.shields.io/github/commits-since/m-o-z-g/marxco-icf/latest/master?include_prereleases&label=commits%20from%20latest%20release)](https://gitlab.com/M-O-Z-G/marxco-icf/master/commits) [![Support the Author](https://frama.link/Wn756FPF)](https://www.blockonomics.co/pay-url/029f3f4400fc11eb) [![Patreon](https://img.shields.io/static/v1?logo=patreon&label=&message=support%20on%20Patreon&color=F96854&logoColor=white)](https://patreon.com/mozgstudio)

## What is?

Marxco Iconfonts provides specific icons for 3D artists, print designers, merchants, and also reflect some cultural artifacts. This icon font is the part of big ambitious Stylus framework Marxco, designed for artists, and with the spirit of classic typesetting in its heart. Which however is in the deep development process, and perhaps will never be released for the public. But it's necessary to have some regular and specific icons under the hand, that's why this project has been separated into own repository, and I hope could be helpful for someone else.

## How to?

All icons are provided in common font formats and ready to use with your code and injected `icons.css` from `public` directory.

### Example:

```html
<i class="mrx_ic mrx_ic-logo-gitlab"></i>
```

> Since 1.7.0 version, source icons related styles remade with Stylus, and now you have full control of prefixes, and its allows to use default variables from Marxco or [Bulma Stylus Plus](https://github.com/M-O-Z-G/bulma-stylus-plus) as default  values, with variable `$framework`. You can use it however you need. Initially, this repo suggested to work as is with CSS.

> Since 1.7.0 version [Gulp](https://gulpjs.com) is doing generation and preparation work only. Showcase layout generated with [Parcel](https://parceljs.org).

But if you're planning to make some changes:

1. You need to install [Node.js](https://nodejs.org/en/).

2. You need to install [Yarn](https://yarnpkg.com/lang/en/).

3. You need to install [Gulp](https://gulpjs.com/).

4. Clone this repository to local.

5. Run terminal, set current dir to the path of your local repository.

6. Run `yarn install` and wait for finished of installation all dependencies.

	Next commands available:
- `npm run build` to optimize, organize new icons then generate fonts and build showcase layout.
- `npm run font-generation` to generate fonts with no processing.
- `gulp test` like `npm run build` but resets all icon unicode names then give new.

### Supported prefixes by default:

`logo-`, `badge-`, `social-`, `finance-`, `market-`, `grafx-`, `3d-`, `doc-`, `software-`, `lic-`, `dev-`, `polit-`, `cult-`, `gem-`, `userpic-`. Any other prefixes and filenames will be proceeded as `miscellaneous` icons.

CHANGELOG.md
# Changelog

## 1.7.0 - 2020-10-12

### Added

- `cult-biohazard`
- `cult-covid-19`
- `cult-jive-belarus`
- `dev-css3-alt`
- `dev-html5`
- `dev-js-deliver`
- `dev-processwire`
- `dev-stylus`
- `dev-wordpress`
- `doc-pdf`
- `finance-blockonomics`
- `finance-buymeacoffee`
- `finance-embily` and `finance-embily-circle`.
- `finance-ko-fi-circle`
- `finance-ko-fi`
- `finance-mastercard`
- `finance-patreon-circle` and `finance-patreon-square`.
- `finance-ttmcard`
- `finance-webmoney`
- `logo-android-2019`
- `logo-creative-cloud`
- `logo-freelancer_com`
- `logo-lets-encrypt`
- `logo-mediafire`
- `logo-mega-nz`
- `logo-openinitiative`
- `logo-origin`
- `market-etsy-square` and `market-etsy`.
- `market-f-droid`
- `social-artstation`
- `social-disroot`
- `social-tiktok`
- `software-hangout`
- `software-viber`
- `userpic-boy`
- `userpic-girl`
- `userpic-man`
- `userpic-women`
- `polit-democratic-party-us`
- `polit-democratic-party-us-modern`
- `polit-republican-party-us`
- `polit-uk`
- `clipboard`
- `copy`
- `download`
- `edit-object`
- `edit`
- `erase`
- `gallery` and `gallery-circles`.
- `list`
- `paste`
- `rss`
- `star`, `star-solid` and `star-half`.
- `upload`


## 1.6.0 - 2020-02-24

### Added

- `finance-liberapay` and `finance-liberapay-square`.
- `social-pawoo` and `social-pawoo-square`.
- `social-deviantart-square`
- `social-flickr-square`
- `social-mastodon-square`
- `social-medium-square`
- `social-steemit` and `steemit-square`.
- `gem-asscher`
- `gem-cushion`
- `gem-emerald`
- `gem-heart`
- `gem-marquise`
- `gem-princess`
- `gem-radiant`
- `doc-afpub`
- `logo-gitlab-square`
- `weight`

### Changed

- Removed loose parts from `gem-oval` and `gem-round`
- `logo-github`
- `logo-gumroad-square`
- `doc-afphoto`
- `doc-afdesign`
- Extra `title` and `defs` tags removed from SVGs.
- Code has been rewritten with ECMAScript 6 to match the requirements of Gulp 4.x and Node.JS 12.
- Icons selectors now are sorted by category and iconname.

## 1.5.0 - 2019-11-04

### Added

- `gems` category in range `62905-63004`.
- `gem-oval`
- `gem-pear`
- `gem-round`
- `cult-skull`
- `cult-jack-o-lantern`


index.js
/*!
* Marxco Iconfont MS
* Icon font from upcoming Marxco component system.
*
* https://gitlab.com/M-O-Z-G/marxco-icf
*
* MIT Licensed
*/

const stylus = require( "stylus" ),
path     = require( "path" ),
data     = require("data.json")
nodes    = stylus.nodes,
utils    = stylus.utils;

exports = module.exports = function() {
	return function( oStyle ){
		oStyle.include( __dirname );
	};
};

Object.defineProperty(icons, "get", {
	enumerable: !1,
	value: function (t) {
		if (icons[t]) return icons[t];
		var c = t.toLowerCase();
		for (var l in icons) {
			var s = icons[l];
			if (s.title && s.title.toLowerCase() === c || s.slug && s.slug === c) return s
		}
	}
}), module.exports = icons;

exports.version = require( path.join( __dirname, "package.json" ) ).version;

exports.path = __dirname;


public/js/main.js
parcelRequire=function(e,t,o,n){function r(o,n){function c(e){return r(c.resolve(e))}function i(t){return e[o][1][t]||t}if(!t[o]){if(!e[o]){var s="function"==typeof parcelRequire&&parcelRequire
if(!n&&s)return s(o,!0)
if(l)return l(o,!0)
if(a&&"string"==typeof o)return a(o)
var d=new Error("Cannot find module '"+o+"'")
throw d.code="MODULE_NOT_FOUND",d}c.resolve=i,c.cache={}
var u=t[o]=new r.Module(o)
e[o][0].call(u.exports,c,u,u.exports,this)}return t[o].exports}function c(e){this.id=e,this.bundle=r,this.exports={}}var l="function"==typeof parcelRequire&&parcelRequire,a="function"==typeof require&&require
r.isParcelRequire=!0,r.Module=c,r.modules=e,r.cache=t,r.parent=l,r.register=function(t,o){e[t]=[function(e,t){t.exports=o},{}]}
for(var i,s=0;s<o.length;s++)try{r(o[s])}catch(d){i||(i=d)}if(o.length){var u=r(o[o.length-1])
"object"==typeof exports&&"undefined"!=typeof module?module.exports=u:"function"==typeof define&&define.amd?define(function(){return u}):n&&(this[n]=u)}if(parcelRequire=r,i)throw i
return r}({bbWE:[function(e,t,o){var n=arguments[3]
Object.defineProperty(o,"__esModule",{value:!0}),o.modal=o.dropdown=o.navbar=void 0
var r=function(e){return Array.prototype.slice.call(document.querySelectorAll(e),0)},c=function(e,t){e.forEach(function(e){e.classList.remove(t)})}
n.$bspSelectors={navbarBurger:".navbar-burger",dropdown:".dropdown",modal:".modal",modalButton:".modal-button",modalBackground:".modal-background",modalClose:".modal-close",modalCardHead:".modal-card-head","delete":".delete",modalCardFoot:".modal-card-foot",button:".button",isClipped:"is-clipped",isHoverable:".is-hoverable",isActive:"is-active"},o.navbar=function(){var e=Array.prototype.slice.call(document.querySelectorAll($bspSelectors.navbarBurger),0)
e.length>0&&e.forEach(function(e){e.addEventListener("click",function(){console.log("Clicked!")
var t=e.dataset.target,o=document.getElementById(t)
e.classList.toggle($bspSelectors.isActive),o.classList.toggle($bspSelectors.isActive)})})},o.dropdown=function(){var e=r($bspSelectors.dropdown+":not("+$bspSelectors.isHoverable+")")
e.length>0&&(e.forEach(function(e){e.addEventListener("click",function(t){t.stopPropagation(),e.classList.toggle($bspSelectors.isActive)})}),document.addEventListener("click",function(t){e.forEach(function(e){e.classList.remove($bspSelectors.isActive)})}))},o.modal=function(){function e(e){var t=document.getElementById(e)
o.classList.add($bspSelectors.isClipped),t.classList.add($bspSelectors.isActive)}function t(){n.forEach(function(e){e.classList.add("is-closing")}),setTimeout(function(){o.classList.remove($bspSelectors.isClipped),n.forEach(function(e){e.classList.remove($bspSelectors.isActive),e.classList.remove("is-closing")})},500)}var o=document.documentElement,n=r($bspSelectors.modal),l=r($bspSelectors.modalButton),a=r($bspSelectors.modalBackground+", "+$bspSelectors.modalClose+", "+$bspSelectors.modalCardHead+" "+$bspSelectors["delete"]+", "+$bspSelectors.modalCardFoot+" "+$bspSelectors.button)
l.length>0&&l.forEach(function(t){t.addEventListener("click",function(){var o=t.dataset.target
t.dataset.slug
e(o)})}),a.length>0&&a.forEach(function(e){e.addEventListener("click",function(){t()})}),document.addEventListener("keydown",function(e){var o=e||window.event
27===o.keyCode&&(t(),c(r($bspSelectors.dropdown+":not("+$bspSelectors.isHoverable+")"),$bspSelectors.isActive))})}},{}],ZCfc:[function(e,t,o){var n=arguments[3],r=this&&this.__createBinding||(Object.create?function(e,t,o,n){void 0===n&&(n=o),Object.defineProperty(e,n,{enumerable:!0,get:function(){return t[o]}})}:function(e,t,o,n){void 0===n&&(n=o),e[n]=t[o]}),c=this&&this.__setModuleDefault||(Object.create?function(e,t){Object.defineProperty(e,"default",{enumerable:!0,value:t})}:function(e,t){e["default"]=t}),l=this&&this.__importStar||function(e){if(e&&e.__esModule)return e
var t={}
if(null!=e)for(var o in e)Object.hasOwnProperty.call(e,o)&&r(t,e,o)
return c(t,e),t}
Object.defineProperty(o,"__esModule",{value:!0})
var a=l(e("./scripts/bulma.ts"))
n.appendData=function(e,t,o,n){console.log("Target is: "+n)
var r=document.querySelector("#"+n),c=r.querySelector("#icDataURI"),l=r.querySelector("#icSVG")
c.value=o,l.value=t,r.querySelector("#icSave").setAttribute("href",o),r.querySelector("#icSave").setAttribute("download",e+".svg"),r.querySelector("figure").innerHTML=t,r.querySelector("#icName").innerHTML=e,c.addEventListener("click",function(){this.select()}),l.addEventListener("click",function(){this.select()}),r.querySelector("#icCopy").addEventListener("click",function(){c.select(),document.execCommand("copy")})}
var i=function(e){return Array.prototype.slice.call(document.querySelectorAll(e),0)}
n.$bspSelectors={navbarBurger:".navbar-burger",dropdown:".dropdown",modal:".modal",modalButton:".modal-button",modalBackground:".modal-background",modalClose:".modal-close",modalCardHead:".modal-card-head","delete":".delete",modalCardFoot:".modal-card-foot",button:".button",isClipped:"is-clipped",isHoverable:".is-hoverable",isActive:"is-active"},document.addEventListener("DOMContentLoaded",function(){var e=document.getElementById("icsb"),t=(i(".icon-wall .columns .column"),document.querySelectorAll(".icon-wall .columns .column")),o=i('.search-boxes input.input[type="search"]')
if("undefined"!=typeof e&&null!=e){var n=e.dataset.string
document.querySelectorAll('.column[id^="ic-"][id*="'+n+'"]')}var r=function(e){$newValue=e.value,o.forEach(function(e){e.value=$newValue}),$allIcons=i(".icon-wall .column"),$matchedIcons=i(".icon-wall .column[data-slug*='"+$newValue+"']"),$allIcons.forEach(function(e){if(e.dataset.slug.search($newValue)>=0)e.style.removeProperty?e.style.removeProperty("display"):e.style.removeAttribute("display")
else if("none"!==window.getComputedStyle(e).display)return void(e.style.display="none")})}
"undefined"!=typeof e&&null!=e&&(console.log("Inputs found: "+o.length),o.forEach(function(e){e.addEventListener("input",function(){return r(e)})})),a.navbar(),a.dropdown(),fetch("data.json").then(function(e){return e.json()}).then(function(e){var o=e
a.modal()
var n=document.documentElement,r=i($bspSelectors.modal),c=i("#icon-preview #closeModal")
c.length>0&&c.forEach(function(e){e.addEventListener("click",function(){r.forEach(function(e){e.classList.add("is-closing")}),window.setTimeout(function(){n.classList.remove($bspSelectors.isClipped),r.forEach(function(e){e.classList.remove($bspSelectors.isActive),e.classList.remove("is-closing")})},500)})}),t.length>0&&t.forEach(function(e){e.addEventListener("click",function(){var t=e.dataset.target,n=e.dataset.slug,r=atob(o[n].dataURI.replace(/^.*,/,"")),c=o[n].dataURI
console.log(o),appendData(n,r,c,t)})})})["catch"](function(e){return console.log(e)})})},{"./scripts/bulma.ts":"bbWE"}]},{},["ZCfc"],null)


gulpfile.js
"use strict";

const gulp         = require('gulp'),
	fs             = require('fs'),
	path           = require('path'),
	pub            = 'public',
	src            = 'src',
	iconsPath     = `${src}/icons`,
	fontPathSrc   = `${src}/font`,
	templatesPath = `${src}/templates`,
	packageJSON   = require( path.join( __dirname, "package.json" ) );

let icRanges      = {
	"logo":          [60905, 61004], //100
	"badge":         [61005, 61104], //100
	"social":        [61105, 61404], //300
	"finance":       [61405, 61504], //100
	"market":        [61505, 61604], //100
	"grafx":         [61605, 61704], //100
	"3d":            [61705, 61804], //100
	"doc":           [61805, 62304], //500
	"software":      [62305, 62404], //100
	"lic":           [62405, 62504], //100
	"dev":           [62505, 62704], //200
	"polit":         [62705, 62804], //100
	"cult":          [62805, 62904], //100
	"gem":           [62905, 63004], //100
	"userpic":       [63005, 63104], //100
	"miscellaneous": [59905, 60904]  //1000
};

const ansiColors     = require('ansi-colors'),
	async            = require('async'),
	consolidate      = require('gulp-consolidate'),
	debug            = require('gulp-debug'),
	del              = require('del'),
	tap              = require('gulp-tap'),
	iconFont         = require('@m-o-z-g/gulp-iconfont'),
	notify           = require('gulp-notify'),
	rename           = require('gulp-rename'),
	sort             = require('gulp-sort'),
	svgo             = require('gulp-svgo'),
	uglify           = require('gulp-uglify'),
	encodeSVGDatauri = require('./node_modules/svgo/lib/svgo/tools.js').encodeSVGDatauri,
	vinylPaths       = require('vinyl-paths');

// Error notificator
function reportError(error) {
	let lineNumber = (error.lineNumber) ? `${LINE} error.lineNumber -- ` : ``;

	notify({
		title: `Task Failed [${error.plugin}]`,
		message: `${lineNumber}See console.`
	}).write(error);

	let report = '';
	let chalk = ansiColors.white.bgRed;

	report += `${chalk('TASK:')} [${error.plugin}]`;
	report += `${chalk('PROB:')} [${error.message}]`;
	if (error.lineNumber) { report += `${chalk('LINE:')} ${error.lineNumber}`; }
	if (error.fileName)   { report += `${chalk('FILE:')} ${error.fileName}`; }
	console.error(report);

	this.emit('end');
};

function checkForUnicodeInFilename(e) {
	if ( e.match(/^u[0-9A-F]{4,5}-/g) != null ) {
		return e.match(/^u[0-9A-F]{4,5}-/g)[0].length;
	}

	return false;
};

function randomHash() {
	let hash = "",
	chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

	for (let i = 0; i < 10; i++) {
		hash += chars.charAt(Math.floor(Math.random() * chars.length));
	}

	return hash;
};

const fileVersion = randomHash();

// SVG Optimizer
function optimize(cb) {
	return gulp
	.src(`${iconsPath}/**/*svg`)
	.pipe(svgo({
		plugins: [{
			cleanupAttrs: true
		}, {
			removeAttrs: {attrs: "(fill-rule|fill|stroke-linejoin|stroke-miterlimit|stroke|clip-rule)"}
		}, {
			addAttributesToSVGElement: {attributes: ["role=\"img\""]}
		}, {
			sortAttrs: true
		}]
	}))
	.pipe(gulp.dest(iconsPath))
	.on('finish', cb);
};

// Clean icons names
function cleanNames(cb) {
	return gulp
		.src(`${iconsPath}/**/*.svg`, {base: './'})
		.pipe(vinylPaths(del))
		.pipe(rename(function(path) {
			let l= checkForUnicodeInFilename(path.basename)
			path.basename= (l != null) ? path.basename.substr(l) : path.basename;
		}))
		.pipe(gulp.dest('./'))
		.on('finish', cb);
};

// Grouping icons within unicode ranges
function unicodify(cb) {
	gulp.src(`${iconsPath}/**/*.svg`, {base: './'})
	.pipe(vinylPaths(del))
	.pipe(rename(function(path) {
		if (checkForUnicodeInFilename(path.basename) == false) {
			let prefix= path.basename.match(/^(.*?)(?=-)/),
			u= '';
		
			if (prefix != null && icRanges.hasOwnProperty(prefix[0])) {
				prefix= prefix[0];
			} else {
				prefix= 'miscellaneous';
			}
		
			let folderName= path.dirname.toString().replace(/.*\\/, '');
		
			if (folderName.match(/^zz_fa/)) {
				u= 'u' + icRanges[prefix][1].toString(16).toUpperCase();
				icRanges[prefix][1] = icRanges[prefix][1] - 1;
			} else {
				u= 'u' + icRanges[prefix][0].toString(16).toUpperCase();
				icRanges[prefix][0] = icRanges[prefix][0] + 1;
			}
			
			if (icRanges[prefix][1] <= icRanges[prefix][0]) {
				notify({
					title: 'Some icons are out of pool!',
					message: 'Icons in `' + prefix + '` group are out of pool! Please, check them!'
				}).write(error);
			}
		
			path.basename = u + '-' + path.basename;
		}
	}))
	.pipe(gulp.dest('./'))
	.on('finish', cb);
};

// Clean
function clean(cb) {
	return del(pub);
};

function organize(done) {
	const licIcons    = `${iconsPath}/lic-*.svg`,
		socialIcons   = `${iconsPath}/social-*.svg`,
		devIcons      = `${iconsPath}/dev-*.svg`,
		financeIcons  = `${iconsPath}/finance-*.svg`,
		logoIcons     = `${iconsPath}/logo-*.svg`,
		politIcons    = `${iconsPath}/polit-*.svg`,
		softwareIcons = `${iconsPath}/software-*.svg`,
		grafxIcons    = `${iconsPath}/grafx-*.svg`,
		threeDeIcons  = `${iconsPath}/3d-*.svg`,
		docIcons      = `${iconsPath}/doc-*.svg`,
		badgeIcons    = `${iconsPath}/badge-*.svg`,
		marketIcons   = `${iconsPath}/market-*.svg`,
		cultIcons     = `${iconsPath}/cult-*.svg`,
		gemIcons      = `${iconsPath}/gem-*.svg`,
		userpicIcons  = `${iconsPath}/userpic-*.svg`,
		 miscellaneousIcons = [
		`${iconsPath}/*.svg`,
		`!${licIcons}`,
		`!${socialIcons}`,
		`!${devIcons}`,
		`!${financeIcons}`,
		`!${logoIcons}`,
		`!${politIcons}`,
		`!${softwareIcons}`,
		`!${grafxIcons}`,
		`!${threeDeIcons}`,
		`!${docIcons}`,
		`!${badgeIcons}`,
		`!${marketIcons}`,
		`!${cultIcons}`,
		`!${gemIcons}`,
		`!${userpicIcons}`
	];

	async.series([
		function (cb) {
			gulp.src(licIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/licenses/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(socialIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/social/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(devIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/dev/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(financeIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/finance/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(logoIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/logotypes/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(politIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/politics/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(softwareIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/software/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(grafxIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/graphics/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(threeDeIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/3d/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(docIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/documents/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(badgeIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/badges/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(marketIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/marketplaces/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(cultIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/culture/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(gemIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/gems/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(userpicIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/userpics/`))
			.on('end', cb);
		},
		function (cb) {
			gulp.src(miscellaneousIcons)
			.pipe(vinylPaths(del))
			.pipe(gulp.dest(`${iconsPath}/miscellaneous/`))
			.on('end', cb);
		}
	],
	function (err, values) {
		if (err) { reportError(err); }
		else { done(); }
	});
}

function generation(done) {
	let fontName = 'Marxco Icons CS',
		prefix = 'mrx_',
		selector = 'ic',
		_tempArray = {},
		iconStream = gulp.src(`${iconsPath}/**/*.svg`)
		.pipe(sort({
			comparator: function(file1, file2) {
				function sanitizeIconName(f) {
					let out = new Object();
					out.name = f.path.replace(/.*src([/\\]|\\\\)icons([/\\]|\\\\).*u[A-F0-9]+-/, '');
					out.categorized = out.name.split('-')[0].match('lic|social|dev|finance|logo|polit|software|grafx|3d|doc|badge|market|cult|gem|userpic');
					out.categorized = (out.categorized != null) ? true : false;
					return out;
				}
				const f1 = sanitizeIconName(file1),
					f2   = sanitizeIconName(file2);
				let f1n = f1.name,
					f2n = f2.name,
					f1c = f1.categorized,
					f2c = f2.categorized;
				if (f1c < f2c) {
					return 1;
				} else if (f1c > f2c) {
					return -1;
				} else {
					if (f1n > f2n) {
						return 1;
					}
					if (f1n < f2n ) {
						return -1;
					}
				}
				return 0;
			}
		}))
		.pipe(tap(function (file, t) {
			let l = checkForUnicodeInFilename(file.stem),
			$fn = (l != null) ? file.stem.substr(l) : file.stem,
			$svg = file.contents.toString();
			_tempArray[$fn] = {
				svg: $svg.replace(/"/g, '\\"'),
				dataURI: encodeSVGDatauri( $svg )
			};
		}))
		.pipe(iconFont({
			fontName: fontName,
			formats: ['svg', 'ttf', 'eot', 'woff', 'woff2'],
			copyright: packageJSON.author.name,
			description: packageJSON.description,
			version: packageJSON.version.replace(/(?<=^([0-9]+)\.([0-9]+)).*/, ''),
			ascent: 896,
			descent: 128,
			url: packageJSON.author.url,
			prependUnicode: true
		}))
		.on('glyphs', function(glyphs, options) {
			for (let i = 0; i < glyphs.length; i++) {
				let glyphName = glyphs[i].name;
				glyphs[i] = { ...glyphs[i], ..._tempArray[glyphName] }
			}
		
			console.log(glyphs, options);
		})

	async.parallel([
		function handleJSON (cb) {
			iconStream.on('glyphs', function(glyphs, options) {
				gulp.src(`${templatesPath}/_ic-font.json`)
				.pipe(consolidate('lodash', {
					glyphs: glyphs
				}))
				.pipe(rename('data.json'))
				.pipe(gulp.dest(src))
				.on('finish', cb);
			});
		},
		function handleIcons (cb) {
			iconStream.on('glyphs', function(glyphs, options) {
				gulp.src(`${templatesPath}/_icons.styl`)
				.pipe(consolidate('lodash', {
					fontName: fontName,
					prefix: prefix,
					selector: selector,
					glyphs: glyphs,
					fontPath: '/font/',
					fileVersion: fileVersion
				}))
				.pipe(rename('icons.styl'))
				.pipe(gulp.dest(src))
				.on('finish', cb);
			});
		},
		function handleShowcase (cb) {
			iconStream.on('glyphs', function(glyphs, options) {
				gulp.src(`${templatesPath}/_icons.pug`)
				.pipe(consolidate('lodash', {
					glyphs: glyphs
				}))
				.pipe(rename('_icons.pug'))
				.pipe(gulp.dest(src + '/layout/dataset/'))
				.on('finish', cb);
			});
		},
		function handleFonts (cb) {
			iconStream
				.pipe(gulp.dest(fontPathSrc))
				.on('finish', cb);
		}
		
	], done);
};

const build = gulp.series(clean, optimize, organize, unicodify, generation);
const test = gulp.series(clean, optimize, cleanNames, organize, unicodify, generation);

// Export tasks
exports.cleanNames = cleanNames;
exports.unicodify = unicodify;
exports.clean = clean;
exports.organize = organize;
exports.generation = generation;
exports.optimize = optimize;
exports.default = build;
exports.test = test;

